utils_test.go

```diff
@@ -42,3 +50,202 @@ func TestValidStoragePathFormat(t *testing.T) {
 		assert.NilError(t, err)
 	}
 }
+
+func TestRootlessRuntimeDir(t *testing.T) {
+	tmpDir, err := ioutil.TempDir(os.TempDir(), "containers-user-*")
+	defer os.RemoveAll(tmpDir)
+	assert.NilError(t, err)
+
+	// testing homedir.GetRuntimeDir binding.
+
+	monkey.Patch(homedir.GetRuntimeDir, func() (string, error) {
+		return tmpDir, nil
+	})
+	defer monkey.Unpatch(homedir.GetRuntimeDir)
+
+	const rootlessUID = math.MaxUint16
+
+	rootlessRuntimeDir, err := getRootlessRuntimeDir(rootlessUID)
+	assert.NilError(t, err)
+	assert.Assert(t, rootlessRuntimeDir == tmpDir, "Rootless runtime dir should follow homedir.GetRuntimeDir.")
+
+	monkey.Patch(homedir.GetRuntimeDir, func() (string, error) {
+		return "", errors.New("homedir.GetRuntimeDir is disabled")
+	})
+
+	// testing systemd dir for systemd based os.
+
+	systemdRuntimeDir := fmt.Sprintf("/run/user/%d", rootlessUID)
+
+	var readFileGuard *monkey.PatchGuard
+	readFileGuard = monkey.Patch(ioutil.ReadFile, func(path string) ([]byte, error) {
+		readFileGuard.Unpatch()
+		defer readFileGuard.Restore()
+		if path == "/proc/1/comm" {
+			return []byte("systemd"), nil
+		}
+		return ioutil.ReadFile(path)
+	})
+	defer monkey.Unpatch(ioutil.ReadFile)
+
+	var lstatGuard *monkey.PatchGuard
+	lstatGuard = monkey.Patch(system.Lstat, func(path string) (*system.StatT, error) {
```

Comment:
For the `ReadFile`/`Lstat` patching, instead `getRootlessRuntimeDir` can be split into a “production” caller with the real paths, and `getRootlessRuntimeDirImplementation` where the relevant paths are parameters, and the tests can point those parameters to test data.

Reply:
Yes, we can split this function and remove several monkey patches.